reactor-core/src/main/java/reactor/core/publisher/Hooks.java

```diff
@@ -655,21 +684,37 @@ static boolean initStaticGlobalTrace() {
 	}
 
 	static <T, P extends Publisher<T>> Publisher<T> addAssemblyInfo(P publisher, AssemblySnapshot stacktrace) {
-		if (publisher instanceof Callable) {
+		if (GLOBAL_STATS_TRACE) {
```

Comment:
I'm not sure I grok all the details of this PR, but shouldn't this be _additive_ to other hooks that were already there?
If it still is with this PR (via code duplication), isn't this problematic in the long run? What if we later add a **third** concern on top of stats collecting? And a fourth, _etc._

Reply:
> I'm not sure I grok all the details of this PR, but shouldn't this be _additive_ to other hooks that were already there?

This feature is *additive* as much as possible. It reuses assembly info in order to collect operators location. It is fully compatible with ReactorDebugAgent which should simplify future users adoption. I'm not still sure that enabling of this future should live separately, but I belive @bsideup can suggest one. I watch thinking about overloaded `onOperatorDebug(boolean enables tats) ` but this will not work with ReactorDebugAgent. Thus the best is a separate hooks method 



> If it still is with this PR (via code duplication). 

Can you please be more certain here? I m not sure about which code duplication your are talking about. 

> , isn't this problematic in the long run? What if we later add a **third** concern on top of stats collecting? And a fourth, _etc._

If you can imagine any extra concerns - all suggestions are welcome. My main goal is to simplify users life. 

Cheers, 
Oleh

Reply:
I guess you are missing a few important points of the full picture. 

1, The main point is that all these methods are internal. 
1. The usage of those internal methods is only within the reactor `onAssembly` or by ReactorDebugAgent. 
1. Those methods clearly state (@Deprecated only for internal usage)
1. From the user perspective, a user has a separate configuration (`enableStatsRecording`) which is SEPARATE from `onOperatorDebug` which enables Stack Trace tracking. These features a coupled. In order to have stats we need operator name and location, so without onOperatorDebug enableStatsReacording should not work. Nevertheless, the implementation of Flux decoration is fully different, so mixing both will be misleading. 

Alright, imagine we follow the suggestions and create a separate method that does the same as `addAssemblyInfo` and call it `addStats`. Now we would need to :

![image](https://user-images.githubusercontent.com/5380167/85710469-f5c4ab80-b6ee-11ea-8cdb-c8501e48f148.png)


1. Add in 6 more places (see image) if-else statement to follow different branches.
1. Every time we would need to  `addAssemblyInfo` -  we would need to maintain this extra if-else branching.
1. The third-party concern would lead to increasing the number of if-else branches in these 6 or more places

------------------------------------------------------------------------------------------


Alright. I guess it is clear that moving that logic from `AddAssemblyInfo` will be even more challenging than support a few if branches in it. I hope we can agree that we should keep the code in this method. 

Anyways, If we wanna code clarity - I can create 2 methods that are being called from `addAssemblyInfo` since all these are assembly info (the first is `addOperatorsStackTraceInfo`  another is `addOperatorStatsInfo`). Should it eliminate the problem being discussed? 

Cheers,
Oleh

Reply:
> Should it eliminate the problem being discussed?

I think that would be a good thing to do, yes. Thanks !

Reply:
fixed